src/stream/src/executor/source/fetch_executor.rs

```diff
@@ -0,0 +1,467 @@
+            self.connector_params.connector_client.clone(),
+            self.actor_ctx.error_suppressor.clone(),
+        )
+    }
+
+    async fn build_state_table_iter(
+        state_store_handler: &SourceStateTableHandler<S>,
+    ) -> StreamExecutorResult<StateTableIter<'_, S>> {
+        Ok(select_all({
+            let mut store_iter_collect =
+                Vec::with_capacity(state_store_handler.state_store.vnodes().len());
+            for vnodes in state_store_handler.state_store.vnodes().iter_vnodes() {
+                store_iter_collect.push(Box::pin(
+                    state_store_handler
+                        .state_store
+                        .iter_row_with_pk_range(
+                            &(Bound::<OwnedRow>::Unbounded, Bound::<OwnedRow>::Unbounded),
+                            vnodes,
+                            PrefetchOptions::new_for_exhaust_iter(),
+                        )
+                        .await?,
+                ))
+            }
+            store_iter_collect
+        }))
+    }
+
+    #[try_stream(ok = Message, error = StreamExecutorError)]
+    async fn into_stream(mut self) {
+        let mut upstream = self.upstream.take().unwrap().execute();
+        let barrier = expect_first_barrier(&mut upstream).await?;
+
+        let mut core = self.stream_source_core.take().unwrap();
+        let mut state_store_handler = core.split_state_store;
+        let mut state_cache = core.state_cache;
+
+        // Build source description from the builder.
+        let source_desc_builder = core.source_desc_builder.take().unwrap();
+
+        let source_desc = source_desc_builder
+            .build()
+            .map_err(StreamExecutorError::connector_error)?;
+
+        // Initialize state table.
+        state_store_handler.init_epoch(barrier.epoch);
+
+        let mut is_datastream_empty = true;
+        let mut stream = StreamReaderWithPause::<true, StreamChunkWithState>::new(
+            upstream,
+            stream::pending().boxed(),
+        );
+
+        if barrier.is_pause_on_startup() {
+            stream.pause_stream();
+        }
+
+        let mut state_table_iter = Self::build_state_table_iter(&state_store_handler).await?;
+
+        // If it is a recovery startup,
+        // there can be file assignments in the state table.
+        // Hence we try to build a reader first.
+        Self::try_replace_with_new_reader(
+            &mut is_datastream_empty,
+            &state_store_handler,
+            &mut state_cache,
+            core.column_ids.clone(),
+            self.build_source_ctx(&source_desc, core.source_id),
+            &source_desc,
+            &mut stream,
+            &mut state_table_iter,
+        )
+        .await?;
+
+        yield Message::Barrier(barrier);
+
+        while let Some(msg) = stream.next().await {
+            match msg {
+                Err(e) => {
+                    tracing::error!("Fetch error: {:?}", e);
+                    todo!()
```

Comment:
please also call rebuild stream here. It should recover from the latest state in cache.